src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,238 @@
+                        emptyGetAccessor,
+                        emptySetAccessor,
+                        root,
+                        context.CancellationToken)));
+        }
+
+        private async Task<Document> ExpandToFullPropertyAsync(
+            Document document, 
+            SyntaxNode property, 
+            SyntaxNode emptyGetAccessor, 
+            SyntaxNode emptySetAccessor, 
+            SyntaxNode root, 
+            CancellationToken cancellationToken)
+        {
+            // Get the symbol representing the property
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+
+            // generate a name for the new field based on naming preferences
+            var rules = await GetNamingRulesAsync(
+                document, 
+                cancellationToken).ConfigureAwait(false);
+            var fieldName = GenerateFieldName(propertySymbol, rules);
+
+            // expand the property and add the field
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var newRoot = await ExpandPropertyAndAddFieldAsync(
+                document, 
+                property, 
+                emptyGetAccessor, 
+                emptySetAccessor, 
+                root, 
+                propertySymbol, 
+                fieldName, 
+                generator, 
+                cancellationToken).ConfigureAwait(false);
+
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var namingStyleOptions = options.GetOption(SimplificationOptions.NamingPreferences);
+            var rules = namingStyleOptions.CreateRules().NamingRules
+                .AddRange(GetDefaultRule(ImmutableArray.Create(new ModifierKind(ModifierKindEnum.IsStatic)), "s_"))
+                .AddRange(GetDefaultRule(ImmutableArray.Create<ModifierKind>(), "_"));
+            return rules;
+        }
+
+        private static string GenerateFieldName(
+            IPropertySymbol property, 
+            ImmutableArray<NamingRule> rules)
+        {
+            var propertyName = property.Name;
+            var fieldName = "";
+            var isStatic = property.IsStatic;
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(
+                    new SymbolKindOrTypeKind(SymbolKind.Field), 
+                    isStatic ? DeclarationModifiers.Static : DeclarationModifiers.None, 
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).Single();
```

Comment:
Is it safe to `Single` here?  Will there always be exactly one here, not more than one?

Reply:
If MakeCompliant always returns a single string... why is it not typed to return a string?

Reply:
@CyrusNajmabadi   MakeCompliant returns IEnumerable<string> and I was using .Single() to convert it to a string.  I have now changed it to .ToString() instead but if there is a better way let me know.

Reply:
Upon closer inspection... it is not returning a single string.  Will fix up my code to handle accordingly :)